lib/grpc/client/connection.ex

```diff
@@ -367,63 +370,114 @@ defmodule GRPC.Client.Connection do
 
     case resolver.resolve(norm_target) do
       {:ok, %{addresses: addresses, service_config: config}} ->
-        lb_policy =
-          cond do
-            is_map(config) and Map.has_key?(config, :load_balancing_policy) ->
-              config.load_balancing_policy
+        build_balanced_state(base_state, addresses, config, lb_policy_opt, norm_opts, adapter)
+
+      {:error, _reason} ->
+        build_direct_state(base_state, norm_target, norm_opts, adapter)
+    end
+  end
 
-            lb_policy_opt ->
-              lb_policy_opt
+  defp resolve_credential(nil, @secure_scheme), do: default_ssl_option()
+  defp resolve_credential(%GRPC.Credential{} = cred, _scheme), do: cred
+  defp resolve_credential(nil, _scheme), do: nil
+  defp resolve_credential(other, _scheme), do: other
 
-            true ->
-              nil
-          end
+  defp validate_adapter_opts!(opts) when is_list(opts), do: :ok
 
-        lb_mod = choose_lb(lb_policy)
-        {:ok, lb_state} = lb_mod.init(addresses: addresses)
+  defp validate_adapter_opts!(_),
+    do: raise(ArgumentError, ":adapter_opts must be a keyword list if present")
 
+  defp build_compressor_list(compressor, accepted) do
+    [compressor | accepted]
+    |> Enum.reject(&is_nil/1)
+    |> Enum.uniq()
+  end
+
+  defp build_balanced_state(base_state, addresses, config, lb_policy_opt, norm_opts, adapter) do
+    lb_policy =
+      cond do
+        is_map(config) and Map.has_key?(config, :load_balancing_policy) ->
+          config.load_balancing_policy
+
+        lb_policy_opt ->
+          lb_policy_opt
+
+        true ->
+          nil
```

Comment:
This seems a bit redundant. Can `lb_policy_opt` be `false`? If not, we can turn this cond into an if